### PR TITLE
Add outlet boundary condition to the matrix free operator

### PR DESCRIPTION
**Pull request by lpsaavedra**
### Description

The `lethe-fluid-matrix-free` application did not support the outlet boundary condition. This is now integrated into the operators. The boundary condition was implemented using the functions created for the weak Dirichlet boundary conditions in a previous PR (#1275) . These functions were generalized and renamed so that they can be used for any boundary condition that requires the evaluation of face terms.

### Testing

There was no test for the outlet BC. A new test has been added for both the matrix-free and the matrix-based application using the mesh `cylinder_close_outlet.msh`:
- `applications_tests/lethe-fluid-matrix-free/cylinder_outlet_bc.prm`
- `applications_tests/lethe-fluid/cylinder_outlet_bc.prm`

### Documentation

A small note has been added to the documentation specifying which boundary conditions are currently not supported with the matrix-free application.

### Checklist (will be removed when merged)
See [this page](https://chaos-polymtl.github.io/lethe/documentation/contributing.html#pull-requests) for more information about the pull request process.

Code related list:
- [x] All in-code documentation related to this PR is up to date (Doxygen format)
- [x] Lethe documentation is up to date
- [x] New feature has unit test(s) (preferred) or application test(s), and restart files are in the generator folder
- [x] The branch is rebased onto master
- [x] Changelog (CHANGELOG.md) is up to date
- [x] Code is indented with indent-all and .prm files (examples and tests) with prm-indent

Pull request related list:
- [x] Labels are applied
- [x] There are at least 2 reviewers (or 1 if small feature) excluding the responsible for the merge
- [x] The PR description is cleaned and ready for merge

### CODE REVIEW

File: include/solvers/fluid_dynamics_matrix_free_operators.h [hunk 1]
```diff
@@ -352,27 +352,22 @@ class NavierStokesOperatorBase : public Subscriptor
...
+   * functions at quadrature points and perform face integrations.
    */
+  template <bool assemble_residual>
   void
-  do_internal_face_integral_range(
-    const MatrixFree<dim, number>               &matrix_free,
-    VectorType                                  &dst,
-    const VectorType                            &src,
-    const std::pair<unsigned int, unsigned int> &range) const;
+  do_face_integral_local(FEFaceIntegrator &integrator) const;
```
Comment (peterrum) [inline include/solvers/fluid_dynamics_matrix_free_operators.h:364]:
```suggestion
  do_boundary_face_integral_local(FEFaceIntegrator &integrator) const;
```

File: include/solvers/fluid_dynamics_matrix_free_operators.h [hunk 2]
```diff
@@ -383,25 +378,29 @@ class NavierStokesOperatorBase : public Subscriptor
    */
   template <bool assemble_residual>
   void
-  do_weak_dirichlet_boundary_range(
+  do_face_integral_range(
```
Comment (peterrum) [inline include/solvers/fluid_dynamics_matrix_free_operators.h:381]:
```suggestion
  do_boundary_face_integral_range(
```

File: include/solvers/fluid_dynamics_matrix_free_operators.h [hunk 3]
```diff
@@ -352,27 +352,22 @@ class NavierStokesOperatorBase : public Subscriptor
...
-   * integral with access to global vectors, i.e., gathering and scattering
-   * values. This is only required for compilation and not needed for our CG
-   * implementation.
+   * @brief Carry out the integration of boundary face integrals.
    *
-   * @param[in] matrix_free Object that contains all data.
-   * @param[in,out] dst Global vector where the final result is added.
-   * @param[in] src Input vector with all values in all cells.
-   * @param[in] range Range of the face batch.
+   * @tparam assemble_residual Flag to assemble the residual or the jacobian.
```
Comment (peterrum) [inline include/solvers/fluid_dynamics_matrix_free_operators.h:357]:
```suggestion
   * @tparam assemble_residual Flag to assemble the residual or the Jacobian.
```

File: include/solvers/fluid_dynamics_matrix_free_operators.h [hunk 4]
```diff
@@ -352,27 +352,22 @@ class NavierStokesOperatorBase : public Subscriptor
...
-    VectorType                                  &dst,
-    const VectorType                            &src,
-    const std::pair<unsigned int, unsigned int> &range) const;
+  do_face_integral_local(FEFaceIntegrator &integrator) const;
 
   /**
    * @brief Loop over all boundary face batches within certain range and perform a face
    * integral with access to global vectors, i.e., gathering and scattering
-   * values. This is used to weakly impose dirichlet boundary conditions.
+   * values. This is used to weakly impose dirichlet boundary conditions or
```
Comment (peterrum) [inline include/solvers/fluid_dynamics_matrix_free_operators.h:369]:
```suggestion
   * values. This is used to weakly impose Dirichlet boundary conditions or
```

File: source/solvers/fluid_dynamics_matrix_free_operators.cc [hunk 5]
```diff
@@ -863,26 +855,34 @@ NavierStokesOperatorBase<dim, number>::
...
+      // information regarding the boundary faces (e.g. too large vectors)
       for (unsigned int face = n_inner_faces;
            face < n_inner_faces + n_boundary_faces;
            ++face)
         {
           face_integrator.reinit(face);
 
+          // We need to read the values for the outlet boundary condition
+          face_integrator.read_dof_values_plain(newton_step);
+          face_integrator.evaluate(EvaluationFlags::values);
```
Comment (peterrum) [inline source/solvers/fluid_dynamics_matrix_free_operators.cc:878]: This is optional. Let's guard it.

File: source/solvers/fluid_dynamics_matrix_free_operators.cc [hunk 6]
```diff
@@ -1164,24 +1182,56 @@ NavierStokesOperatorBase<dim, number>::do_local_weak_dirichlet_bc(
...
+                value_result[d] -=
+                  kinematic_viscosity * gradient[d][i] * normal_vector[i];
+              // Assemble ν(∇v·n,(u-u_target))
+              for (unsigned int i = 0; i < dim; ++i)
+                gradient_result[d][i] -=
+                  kinematic_viscosity * value[d] * normal_vector[i];
+            }
+        }
+      else if (this->boundary_conditions.type[boundary_index] ==
+               BoundaryConditions::BoundaryType::outlet)
```
Comment (peterrum) [inline source/solvers/fluid_dynamics_matrix_free_operators.cc:1209]: I would restructure the code a bit, since gradients are not needed in this code path.

Comment (lpsaavedra) [reply source/solvers/fluid_dynamics_matrix_free_operators.cc:1209]: True, I will refactor the code a bit to make sure that we only do what the specific boundary condition requires

File: source/solvers/fluid_dynamics_matrix_free_operators.cc [hunk 7]
```diff
@@ -1151,44 +1174,90 @@ NavierStokesOperatorBase<dim, number>::do_local_weak_dirichlet_bc(
...
+
+
+          // If we are assembling the residual, substract the target velocity
+          // from the velocity value.
+          if constexpr (assemble_residual)
+            for (unsigned int d = 0; d < dim; ++d)
+              value[d] -= this->face_target_velocity[face_index][q][d];
+
+          for (unsigned int d = 0; d < dim; ++d)
+            { // Assemble (v,beta (u-u_target)) for the main penalization
```
Comment (peterrum) [inline source/solvers/fluid_dynamics_matrix_free_operators.cc:1201]:
```suggestion
            { 
            // Assemble (v,beta (u-u_target)) for the main penalization
```

**Review by peterrum — APPROVED** [review]

File: source/solvers/fluid_dynamics_matrix_free_operators.cc [hunk 8]
```diff
@@ -1151,44 +1174,90 @@ NavierStokesOperatorBase<dim, number>::do_local_weak_dirichlet_bc(
...
+
+          auto       value    = integrator.get_value(q);
+          const auto gradient = integrator.get_gradient(q);
+
+
+          // If we are assembling the residual, substract the target velocity
+          // from the velocity value.
+          if constexpr (assemble_residual)
+            for (unsigned int d = 0; d < dim; ++d)
+              value[d] -= this->face_target_velocity[face_index][q][d];
```
[the remark this answers is not in the document]
Comment (lpsaavedra) [reply source/solvers/fluid_dynamics_matrix_free_operators.cc:1198]: This won't work, because the `face_target_velocity` has only velocity values (dim) and `value` includes also the pressure (dim + 1).  Is there any advantage on doing it like this? If yes, I can simply change the size and add zero for the pressure... however, I find it easier to understand like this...

[the remark this answers is not in the document]
Comment (blaisb) [reply source/solvers/fluid_dynamics_matrix_free_operators.cc:1198]: If you think it's better to keep the loop version instead of having this zero padded with a non-existing pressure, it's good for me.

**Comment by peterrum** [comment]
Good to go once you have added a test!

**Comment by lpsaavedra** [comment]
I have added now a test with the help of @blaisb. It was added for both the matrix-free and matrix-based applications.

**Review by blaisb — APPROVED** [review]
Great work :)
